pkg/northbound/gnmi/subscribe.go

```diff
@@ -15,36 +15,56 @@
 package gnmi
 
 import (
+	"crypto/sha1"
+	"fmt"
 	"github.com/onosproject/onos-config/pkg/events"
 	"github.com/onosproject/onos-config/pkg/manager"
 	"github.com/onosproject/onos-config/pkg/store/change"
 	"github.com/onosproject/onos-config/pkg/utils"
 	"github.com/openconfig/gnmi/proto/gnmi"
 	"io"
 	"log"
-	"strings"
 	"time"
 )
 
-//per each subscribe request we receive the map is updated with a channel corresponding to the path.
-var (
-	PathToChannels = make(map[*gnmi.Path]chan *gnmi.Update)
-)
-
 // Subscribe implements gNMI Subscribe
 func (s *Server) Subscribe(stream gnmi.GNMI_SubscribeServer) error {
-	updateChan := make(chan *gnmi.Update)
+	//updateChan := make(chan *gnmi.Update)
 	var subscribe *gnmi.SubscriptionList
+	mgr := manager.GetManager()
+	//hashing the stream to obtain a unique identifier of the client
+	h := sha1.New()
+	_, err1 := io.WriteString(h, fmt.Sprintf("%v", stream))
+	if err1 != nil {
+		return err1
+	}
+	hash := string(h.Sum(nil))
+	//Registering one listener per NB app/client on both change and opStateChan
+	changesChan, err := mgr.Dispatcher.RegisterNbi(hash)
+	if err != nil {
+		log.Println("Subscription present: ", err)
+		return err
+	}
+	opStateChan, err := mgr.Dispatcher.RegisterOpState(hash)
+	if err != nil {
+		log.Println("Subscription present: ", err)
+		return err
+	}
 	//this for loop handles each subscribe request coming into the server
 	for {
```

Comment:
ok - on the `for` loop - i think I get it now
It doesn't actually loop - it just blocks so that the other threads can send the response back
Maybe change the comment

Reply:
done